driver-core/src/main/java/com/datastax/driver/core/NettyOptions.java

```diff
@@ -0,0 +1,183 @@
+     * do the following:
+     *
+     * <pre>
+     * ByteBufAllocator myCustomByteBufAllocator = ...
+     * boostrap.option(ChannelOption.ALLOCATOR, myCustomByteBufAllocator);
+     * </pre>
+     *
+     * @param bootstrap the {@link Bootstrap} being initialized.
+     */
+    public void afterBootstrapInitialized(Bootstrap bootstrap) {
+        //noop
+    }
+
+    /**
+     * Hook invoked each time the driver creates a new {@link Connection}
+     * and initializes the {@link SocketChannel channel}.
+     * <p>
+     * This hook is guaranteed to be called <em>after</em> the driver has registered
+     * all its internal channel handlers, and applied the configured {@link SSLOptions}, if any.
+     * <p>
+     * This is a good place to add extra {@link io.netty.channel.ChannelHandler ChannelHandler}s
+     * to the channel' pipeline; e.g. to add a custom SSL handler to the beginning of the handler chain,
+     * do the following:
+     *
+     * <pre>
+     * ChannelPipeline pipeline = channel.pipeline();
+     * SSLEngine myCustomSSLEngine = ...
+     * SslHandler myCustomSSLHandler = new SslHandler(myCustomSSLEngine);
+     * pipeline.addFirst("ssl", myCustomSSLHandler);
+     * </pre>
+     *
+     * Note: if you intend to provide your own SSL implementation,
+     * do not enable the driver's built-in {@link SSLOptions} at the same time.
+     *
+     * @param channel the {@link SocketChannel} instance, after being initialized by the driver.
+     * @throws Exception if this methods encounters any errors.
+     */
+    public void afterChannelInitialized(SocketChannel channel) throws Exception {
+        //noop
+    }
+
+    /**
+     * Hook invoked when the cluster is shutting down after a call to {@link Cluster#close()}.
+     * <p>
+     * This is guaranteed to be called only after all connections have been individually
+     * closed, and only once per {@link EventLoopGroup} instance.
+     * <p>
+     * This gives the implementor a chance to close the {@link EventLoopGroup} properly.
+     * The default implementation simply calls {@link EventLoopGroup#shutdownGracefully()} on the passed instance,
+     * then waits uninterruptibly until it is done.
+     * <p>
+     * Implementation note: if the {@link EventLoopGroup} instance is being shared, or used for other purposes than to
+     * coordinate Netty events for the current cluster, than it should not be shut down here;
+     * subclasses would have to override this method accordingly to take the appropriate action.
+     *
+     * @param eventLoopGroup the event loop group used by the cluster being closed
+     */
+    public void onClusterClose(EventLoopGroup eventLoopGroup) {
+        // This will shut down I/O threads. Since this is called only once all connections have
+        // been individually closed, it's fine.
+        eventLoopGroup.shutdownGracefully().syncUninterruptibly();
```

Comment:
One slight regression I just noticed that was introduced in #301 is that it now takes at least 2 seconds now to do Cluster.close() because shutdownGracefully with default parameters employs a quiet period of 2 seconds where previously closing a cluster instance was very quick (milliseconds).  Since we close the channel group before shutting down the eventLoopGroup, I think it would be safe to run this with a grace period of 0 seconds, it would look like this:

```java
eventLoopGroup.shutdownGracefully(0, 15, TimeUnit.SECONDS).syncUninterruptibly();
```

You can see an example of this in our CI results here:
http://cassci.datastax.com/view/y_Drivers/job/y_java-driver_2.0/294/testReport/com.datastax.driver.core/SessionTest/closeDuringClusterInitTest/  (first succeeding commit after java622 merged, takes 18 minutes)

Previously the same test took 13 seconds:
http://cassci.datastax.com/view/y_Drivers/job/y_java-driver_2.0/292/testReport/com.datastax.driver.core/SessionTest/closeDuringClusterInitTest/